### PR TITLE
Fixing src/Lint.jl lintserver conn to close properly and testing it in test/server.jl

### PR DESCRIPTION
**Pull request by TeroFrondelius**
Thanks for the good work. I wrote [an Atom package linter-julia](https://github.com/TeroFrondelius/linter-julia), which uses the lintserver(). However, node.js client thinks lintserver isn't closing properly and thus stay busy. Eventually the fix could have been just the close(conn), but I wanted to keep a backward compatibility thus making this `readandwritethestream()` function of the old `@async try` internals solved the trick by itself. At least those two tests, which I added now works with julia 0.5.0.

### CODE REVIEW

**Comment by TotalVerb** [comment]
We can't use `@testset` unless we do https://github.com/tonyhffong/Lint.jl/issues/174. @tonyhffong @Michael-Klassen thoughts on dropping 0.4, as proposed, so we need not worry about these cases in the future?

**Comment by TeroFrondelius** [comment]
For me it's not a big deal to remove the `@testset` and push again.

File: src/Lint.jl [hunk 1]
```diff
@@ -345,31 +345,36 @@ function lintdir{T<:AbstractString}(dir::T, ctx::LintContext=LintContext())
...
-                msgs = lintfile(file, code)
-                # Write response to socket
-                for i in msgs
-                    write(conn, string(i))
-                    write(conn, "\n")
-                end
-                # Blank line to indicate end of messages
-                write(conn, "\n")
-                println("Connection closed")
+              readandwritethestream(conn)
```
Comment (TotalVerb) [inline src/Lint.jl]: indentation looks off

File: src/Lint.jl [hunk 2]
```diff
@@ -345,31 +345,36 @@ function lintdir{T<:AbstractString}(dir::T, ctx::LintContext=LintContext())
     ctx.messages
 end
 
+function readandwritethestream(conn)
+    println("Connection accepted")
```
Comment (TotalVerb) [inline src/Lint.jl]: not related to this PR, but debugging output should be printed to STDERR

File: src/Lint.jl [hunk 3]
```diff
@@ -345,31 +345,36 @@ function lintdir{T<:AbstractString}(dir::T, ctx::LintContext=LintContext())
...
+    # Get file, code length and code
+    file = strip(readline(conn))
+    println("file: ", file)
+    code_len = parse(Int, strip(readline(conn)))
```
Comment (TotalVerb) [inline src/Lint.jl]: the `strip` is not needed here (not this PR's fault)

File: src/Lint.jl [hunk 4]
```diff
@@ -345,31 +345,36 @@ function lintdir{T<:AbstractString}(dir::T, ctx::LintContext=LintContext())
...
+    msgs = lintfile(file, code)
+    # Write response to socket
+    for i in msgs
+        write(conn, string(i))
+        write(conn, "\n")
+    end
+    # Blank line to indicate end of messages
+    write(conn, "\n")
+    println("Connection closed")
+    close(conn)
```
Comment (TotalVerb) [inline src/Lint.jl]: it seems weird to me to close the connection in this function.  I think it's better to close it in a `finally` in the `lintserver` function.

File: src/Lint.jl [hunk 5]
```diff
@@ -345,31 +345,36 @@ function lintdir{T<:AbstractString}(dir::T, ctx::LintContext=LintContext())
...
-                for i in msgs
-                    write(conn, string(i))
-                    write(conn, "\n")
-                end
-                # Blank line to indicate end of messages
-                write(conn, "\n")
-                println("Connection closed")
+              readandwritethestream(conn)
             catch err
               println("connection ended with error $err")
```
Comment (TotalVerb) [inline src/Lint.jl]: as in review above, I think it's best to close the connection in a `finally` clause here.

File: src/Lint.jl (hunk 4 above)
Comment (TeroFrondelius) [inline src/Lint.jl]: Sorry, I didn't notice your comments earlier. This is the whole point of this pull request. Unless this connection is properly closed the coffeescript code keeps hanging. I don't have the expertise on server client coding, but this is based on observing what is happening in practice. The same happens with tests/server.jl if you comment out the close(conn) here, which gives you possibility to test it yourself easily.

File: src/Lint.jl (hunk 2 above)
Comment (TeroFrondelius) [inline src/Lint.jl]: I will fix this.

File: src/Lint.jl (hunk 1 above)
Comment (TeroFrondelius) [inline src/Lint.jl]: I will fix this. CoffeeScript uses two spaces for indentation and I've been coding that at the same time.

File: src/Lint.jl (hunk 4 above)
Comment (TotalVerb) [inline src/Lint.jl]: I understand. But this is the wrong place to close it. Closing the connection should be done in the try-catch-finally below.

Comment (TotalVerb) [inline src/Lint.jl]: As a rule of thumb, resources should be closed/freed in the same scope as they are allocated / acquired, whenever possible. As it stands, if some error occurs somewhere, the connection will not be closed.

Comment (TeroFrondelius) [inline src/Lint.jl]: Sorry I misunderstood your proposal earlier. Now it makes sense. You proposed this:
```julia
function lintserver(port)
    server = listen(port)
    try
        println("Server running on port $port ...")
        while true
            conn = accept(server)
            @async try
                readandwritethestream(conn)
            catch err
                println("connection ended with error $err")
            finally
                close(conn)
                println("Connection closed")
            end
        end
    finally
        close(server)
    end
end
```
and this works. I will make all the changes and push the commit(s).

File: src/Lint.jl (hunk 2 above)
Comment (TeroFrondelius) [inline src/Lint.jl]: I looked around the sources and the style seems to be: comment out the debugging println() statements. Should I comment out these as well?

Comment (TotalVerb) [inline src/Lint.jl]: This is frankly a bad style. We should switch to a real logging system in the long run. For now anything's fine.

File: src/Lint.jl (hunk 3 above)
Comment (TeroFrondelius) [inline src/Lint.jl]: I tried to remove the first strip() and it did break the backward compatibility:
```bash
Server running on port 2228 ...
lintserver() tests: Test Failed
  Expression: contains(readline(conn),"use of undeclared symbol")
 in record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:428
 in do_test(::Base.Test.Returned, ::Expr) at ./test.jl:281
 in macro expansion; at /home/tero/dev/Lint.jl/test/server.jl:25 [inlined]
 in macro expansion; at ./test.jl:672 [inlined]
 in anonymous at ./<missing>:?
 in include_from_node1(::String) at ./loading.jl:488
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
lintserver() tests: Test Failed
  Expression: readline(conn) == "\n"
   Evaluated: ":1 E321 bad: use of undeclared symbol\n" == "\n"
 in record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:428
 in do_test(::Base.Test.Returned, ::Expr) at ./test.jl:281
 in macro expansion; at /home/tero/dev/Lint.jl/test/server.jl:26 [inlined]
 in macro expansion; at ./test.jl:672 [inlined]
 in anonymous at ./<missing>:?
 in include_from_node1(::String) at ./loading.jl:488
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
Test Summary:      | Pass  Fail  Total
  lintserver() tests |    1     2      3
ERROR: LoadError: Some tests did not pass: 1 passed, 2 failed, 0 errored, 0 broken.
 in finish(::Base.Test.DefaultTestSet) at ./test.jl:495
 in macro expansion; at ./test.jl:679 [inlined]
 in anonymous at ./<missing>:?
 in include_from_node1(::String) at ./loading.jl:488
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
while loading /home/tero/dev/Lint.jl/test/server.jl, in expression starting on line 11
```

Comment (TotalVerb) [inline src/Lint.jl]: Very well, then best to leave it.

**Comment by TeroFrondelius** [comment]
I am just  learning the git and github. It looks like I messed up the committing history. Should I try to fix it somehow or can you track my changes? End result should be ok, if you will do squash merge.

File: src/Lint.jl [hunk 6]
```diff
@@ -1,4 +1,4 @@
-__precompile__(true)
+  __precompile__(true)
```
Comment (TotalVerb) [inline src/Lint.jl]: extra spaces

File: test/server.jl [hunk 7]
```diff
@@ -4,24 +4,57 @@ close(conn[2])
 port = conn[1]
 
 server = @async lintserver(port)
-sleep(1) #let sever start
+sleep(1) #let server start
 
+@testset "lintserver() tests" begin
+  conn = connect(port)
```
Comment (TotalVerb) [inline test/server.jl]: 4 space indent

**Comment by TotalVerb** [comment]
I don't have commit access to this repo, so you'll have to ping @Michael-Klassen or @tonyhffong to get it merged.

**Comment by Michael-Klassen** [comment]
@TotalVerb Tell me when you are satisfied with it and I'll merge it.

File: src/Lint.jl (hunk 6 above)
Comment (TotalVerb) [inline src/Lint.jl]: bump, just in case you missed this comment

Comment (TeroFrondelius) [inline src/Lint.jl]: Yes, I missed. I must have typed that by accident. I remove it.

**Pull request merged**